src/sql/azdata.d.ts

```diff
@@ -189,6 +189,40 @@ declare module 'azdata' {
 		 * @param connectionProfile connection profile
 		 */
 		export function connect(connectionProfile: IConnectionProfile, saveConnection?: boolean, showDashboard?: boolean): Thenable<ConnectionResult>;
+
+		/**
+		 * Supported connection event types
+		 */
+		export type ConnectionEventType =
+			| 'onConnect'
+			| 'onDisconnect'
+			| 'onConnectionChanged';
+
+		/**
+		 * Connection Event Lister
+		 */
+		export interface ConnectionEventListener {
+			/**
+			 * Connection event handler
+			 * @param type Connection event type
+			 * @param ownerUri Connection's owner uri
+			 * @param args Connection profile
+			 */
+			onConnectionEvent(type: ConnectionEventType, ownerUri: string, args: IConnectionProfile): void;
+		}
+
+		/**
+		 * Register a connection event listener
+		 * @param listener The connection event listener
+		 */
+		export function registerConnectionEventListener(listener: connection.ConnectionEventListener): void;
```

Comment:
How do we dispose of this?

Reply:
The extension adds all disposables to its extension context's disposables

Reply:
Yeah but the listener isn't disposable. Amir is right - we should be following what VS code does when you register something where it returns a disposable that when disposed will dispose the listener.

https://github.com/microsoft/azuredatastudio/blob/main/src/vs/vscode.d.ts#L7527

Reply:
I looked at the code a tiny bit and didn't see it doing this, do you have a link to the applicable part?

Reply:
You are right but all our other methods have been void. Do we want to change this paradigm now?

Reply:
I mean its technical debt and since we're introducing this to stable we should at the very least make sure we don't introduce more technical debt to the code.

Listener leaks happen quite often in ADS and we need to start cleaning them up somewhere.

Reply:
Just because something was done wrong in the past doesn't mean we shouldn't do the right thing going forward 😄 And then we should look into modifying the existing APIs too best we can.

Reply:
I agree but this looks like something that should be done in a "clean-up" PR to make the ADS API as uniform as possible. If I were an extension author, I would be pretty confused about the disparity in the API. Either it should stay in the proposed until all of the previous methods are `Disposable` or should be included as `void` for now to have the API uniform. 
The API change will be a breaking one and so it won't be easy, and I don't see it happening in near future. What starting this would do is make the API have even more calls with `Disposables`, only causing confusion. Sure, I don't mind what's done here, but personally, I would really dislike this API as an author myself.

Reply:
Well, we shouldn't be promoting api's into stable that have known bad patterns.

As for the confusion - I don't really think that's a concern. Worst case they don't do anything with the disposable and then it behaves like it did previously. 

Ideally yes, we'd have everything be consistent. But as you mentioned this isn't something we can fix easily in the existing api without more consideration and review so for now it's better to just start doing the right thing and then if people ask explain why the difference exists and our plans going forward.